module/executiondatasync/optimistic_sync/execution_result/info_provider.go

```diff
@@ -106,12 +125,65 @@ func (e *Provider) ExecutionResultInfo(
 		return nil, fmt.Errorf("no execution nodes found for result %v (blockID: %v): %w", result.ID(), blockID, err)
 	}
 
+	header, err := e.headers.ByBlockID(blockID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get header by block ID %v: %w", blockID, err)
+	}
+	// Lookup the finalized block ID at the height of the requested block
+	blockIDFinalized, err := e.headers.BlockIDByHeight(header.Height)
+	if err != nil {
+		return nil, fmt.Errorf("failed to lookup finalized block ID at height %d: %w", header.Height, err)
+	}
+	// If the requested block conflicts with finalized block, return error
+	if blockIDFinalized != blockID {
+		return nil, optimistic_sync.NewBlockFinalityMismatchError(blockID, blockIDFinalized)
+	}
+
 	return &optimistic_sync.ExecutionResultInfo{
 		ExecutionResultID: result.ID(),
 		ExecutionNodes:    subsetENs,
 	}, nil
 }
 
+// validateCriteria verifies that the provided optimistic sync criteria can be
+// satisfied by the currently available execution nodes.
+//
+// The validation ensures that the required executor IDs do not exceed the number
+// of available executors, that the requested AgreeingExecutorsCount is feasible,
+// and that every required executor ID is present in the available set.
+//
+// Expected errors during normal operations:
+//   - [optimistic_sync.RequiredExecutorsCountExceededError]: Required executor IDs count exceeds available executors.
+//   - [optimistic_sync.AgreeingExecutorsCountExceededError]: Agreeing executors count exceeds available executors.
+//   - [optimistic_sync.UnknownRequiredExecutorError]: A required executor ID is not in the available set.
+func (e *Provider) validateCriteria(
+	criteria optimistic_sync.Criteria,
+	availableExecutors flow.IdentityList,
+) error {
+	requiredExecutors := criteria.RequiredExecutors
+	if len(availableExecutors) < len(requiredExecutors) {
```

Comment:
Not all `RequiredExecutors` must produce the result, just at least one:
https://github.com/onflow/flow-go/blob/0f22a33a62e6fb8735272ac151fd455fb53ec23e/module/executiondatasync/optimistic_sync/execution_result_query_provider.go#L11